### ISSUE TITLE
Blazor Web Assembly AAD Auth under .Net 5.0 cannot request multiple tokens

### ISSUE THREAD

**Issue by snapfisher** [dotnet/aspnetcore#29384]
### Describe the bug
I see this was reported in #28226, but it was closed without resolution, and the problem definitely exists.

Following instructions here: https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/additional-scenarios?view=aspnetcore-5.0#request-additional-access-tokens

You cannot request multiple tokens under .Net 5.0.  The error received is: 

Provided value for the input parameter scope is not valid because it contains more than one resource. Scope x y is not valid, where x and y are the scopes I am requesting.

If I request 1 scope only, I can authenticate with no errors.

I have an older blazor wasm application built against .Net Standard 2.1, which follows the same coding instructions/pattern where this does work.

### To Reproduce

Program.cs is:

        public static async Task Main(string[] args)
        {
            var builder = WebAssemblyHostBuilder.CreateDefault(args);
            builder.RootComponents.Add<App>("#app");

            builder.Services.AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) });

            builder.Services.AddMsalAuthentication(options =>
            {
                builder.Configuration.Bind("AzureAd", options.ProviderOptions.Authentication);
                options.ProviderOptions.DefaultAccessTokenScopes.Add("https://graph.microsoft.com/openid");
                options.ProviderOptions.AdditionalScopesToConsent.Add("https://management.azure.com/user_impersonation");
            });

            builder.Services.AddAuthorizationCore(options =>
            {
                options.AddPolicy("OneUser", policy => policy.RequireAssertion(context =>
                    context.User.HasClaim(c =>
                    c.Type == "preferred_username" &&
                    c.Value == "x*x.onmicrosoft.com")
                ));
            });

            await builder.Build().RunAsync();
        }

It fails even without attempting token access.  Index.razor does not really do much, only ...

<AuthorizeView Policy="OneUser">
    <Authorized>
    </Authorized>
    <Authorizing>
        <h1>Authentication in progress</h1>
    </Authorizing>
    <NotAuthorized>
        <h2>Please log in with an authorized user</h2>
    </NotAuthorized>
</AuthorizeView>

### Exceptions (if any)

The error/exception number is AADSTS28000

### Further technical details
- ASP.NET Core version
- Include the output of `dotnet --info`
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version
.NET SDK (reflecting any global.json):
 Version:   5.0.102
 Commit:    71365b4d42

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.102\

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  1.1.13 [C:\Program Files\dotnet\sdk]
  1.1.14 [C:\Program Files\dotnet\sdk]
  2.1.617 [C:\Program Files\dotnet\sdk]
  2.1.700 [C:\Program Files\dotnet\sdk]
  2.1.701 [C:\Program Files\dotnet\sdk]
  2.1.812 [C:\Program Files\dotnet\sdk]
  2.2.300 [C:\Program Files\dotnet\sdk]
  3.1.300 [C:\Program Files\dotnet\sdk]
  3.1.405 [C:\Program Files\dotnet\sdk]
  5.0.102 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

Visual Studio Enterprise 16.8.4

**Comment by snapfisher** [comment]
The AUTHENTICATION IN PROGRESS, was actually supposed to be xaml code.  The markup is too smart for my own good

I added the quotes.....

"<AuthorizeView Policy="OneUser">
    <Authorized>
    </Authorized>
    <Authorizing>
        <h1>Authentication in progress</h1>
    </Authorizing>
    <NotAuthorized>
        <h2>Please log in with an authorized user</h2>
    </NotAuthorized>
</AuthorizeView>"

**Comment by snapfisher** [comment]
The authentication in progress section of the above, was supposed to be the xaml code from the razor page, the markup is to smart for it's own good.  I changed the < characters to # just so it would make it into the issue:


#AuthorizeView Policy="OneUser">
    #Authorized>
    #/Authorized>
    #Authorizing>
        #h1>Authentication in progress</h1>
    #/Authorizing>
    #NotAuthorized>
        #h2>Please log in with an authorized user</h2>
    #/NotAuthorized>
#/AuthorizeView>

**Comment by javiercn** [comment]
@snapfisher thanks for the details.

Seems like something is going on here.

**Comment by BenWhite27** [comment]
This is blocking me at the moment, my app needs to access it's own API and the MS Graph API.  Any workarounds?

**Comment by BenWhite27** [comment]
@SteveSandersonMS I believe the tags could be reviewed for this issue.  Using Microsoft Identity Platform + Graph API for things like avatar images, directory listings, and user management must be a fairly common and documented scenario?

I've started comparing the code provided provided as a resolution to the previous issue (#28226) and trimming it down to the bare essentials.  It appears all that's required to get this working for me with the Graph API is to not specify the scopes when adding the graph client to the services collection.  This in turn doesn't add anything extra to `AdditionalScopesToConsent`.

program.cs in wasm project file.
```diff
- builder.Services.AddGraphClient("https://graph.microsoft.com/User.Read");
+ builder.Services.AddGraphClient();
```
The required scopes are also specified in GraphClientExtensions.cs from the docs to authenticate the request.

I don't understand the inner workings here enough to say if this is going to work moving forward or bite me later in development but might help someone.

**Comment by igalfsg** [comment]
@javiercn @SteveSandersonMS is there any update to this?

**Comment by SteveSandersonMS** [comment]
At the time, we had heard very few reports. More are emerging now so I’m updating the label.

**Comment by jelard** [comment]
This is also a blocker for me.

**Comment by SteveSandersonMS** [comment]
Re-triaging.

**Comment by javiercn** [comment]
I've spent a chunk of time investigating this in the past couple of days and I believe it comes down to some changes AAD made to avoid requesting all scopes upfront. This now needs to be done via incremental consent, which is doable, but requires a different approach.

I'm going to spend some time to see if I can put an example that accomplishes this through our existing extensibility, but in essence, I believe that requesting the scopes upfront does no longer work.

We'll have to revisit the area in .NET 7.0

**Comment by igalfsg** [comment]
@javiercn I don't know if this helps, but how we went around this is creating a different client for each endpoint for example a graph http client, and an arm http client and we use whichever we need depending on the case, maybe doing that but abstracting it from the user so the user only sees one (or use the same client and just attach different tokens based on the scope being contacted).

**Comment by igalfsg** [comment]
Yeah the additional popup is annoying, if the user only has one account SSO takes care of it  but if they have multiple AAD accounts they would get a popup for each scope. 

Here is a GitHub repo : https://github.com/coding-flamingo/BlazorWasmWithAADAuth
and here is the youtube video of me explaining what I did https://www.youtube.com/watch?v=Aa8QTlyNDBM

**Comment by javiercn** [comment]
We addressed this for 7.0, the fix will be available in RC2

https://github.com/dotnet/aspnetcore/pull/43954

**Comment by igalfsg** [comment]
Thanks @javiercn, Will it work out of the box or is there something we have to do?
